### PR TITLE
fix(str): make replace NUL-safe (#1048)

### LINKED ISSUE #1048: fix(str): make `replace` NUL-safe
## Context

#1022 added `StringHeader` with explicit `byte_len`. `replace(s, old, new)` currently uses C `strstr` to find occurrences of `old`, which stops at the first NUL byte.

## Reproduction

```ry
s = "a\0b\0a"
print(replace(s, "\0", "-"))  # Expected: "a-b-a", Actual: "a"
print(replace(s, "a", "X"))   # Expected: "X\0b\0X", Actual: "X"
```

## Fix

Implement NUL-safe search loop using `stringByteLen` + `memmem` (or manual sliding window), then build the result via `makeStringUninit` + `memcpy`.

## Files

- `src/codegen_call_string.cpp` — `emitReplace` (or corresponding runtime function)

## Related

- #1022 (StringHeader introduction)

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

- Replaces the 120-line two-pass IR loop in `emitStrOp_replace` (which used `strstr`/`strlen`, both of which truncate at embedded NUL bytes) with a single call to a new runtime helper `__ry_str_replace`
- `__ry_str_replace` uses `memmem` throughout — haystack, needle, and replacement all support embedded `\0`
- Empty needle returns a fresh copy unchanged (preserves #802 behaviour; guard moved to C++ runtime)
- `count == 0` fast-path avoids an unnecessary allocation when the needle is not found
- Overflow guard prevents int64_t UB when `count * (newLen - oldLen)` would exceed `INT64_MAX`

Closes #1048

## Test plan

- [ ] New `describe("NUL byte handling")` cases in `tests/spec/strings.test.ry` (5 `it` blocks tagged `#1048`):
  - NUL as needle (`replace("a\0b\0a", "\0", "-")` → `"a-b-a"`)
  - NUL in haystack tail (`replace("a\0b\0a", "a", "X")` → `"X\0b\0X"`)
  - Multi-byte needle containing NUL
  - Empty-needle parity
  - NUL bytes in replacement string
- [ ] Existing `replace` tests (lines 71–81) continue to pass
- [ ] `./build/ry_tests` — 1709 C++ tests pass
- [ ] `./build/ry test -p` — 131 Ry self-tests pass
- [ ] ASan + UBSan clean
- [ ] TSan `ry_tests` clean

## Summary by CodeRabbit

* **Bug Fixes**
  * The `replace` string operation now correctly handles and preserves embedded NUL bytes in strings, instead of truncating at the first `\0`.

* **Documentation**
  * Updated reference documentation with examples demonstrating NUL-safe string replacement behavior.

* **Tests**
  * Added test cases for NUL byte handling in the `replace` function, covering edge cases with embedded NUL bytes.

### CODE REVIEW

File: tests/spec/strings.test.ry [hunk 1]
```diff
@@ -397,4 +397,27 @@ describe("NUL byte handling", ():
...
+    expect(replace("xx\0yy\0xx", "\0y", "Z")).to_eq("xxZy\0xx")
+  )
+
+  it("replace empty needle returns input unchanged (NUL parity #1048)", ():
+    expect(replace("a\0b", "", "X")).to_eq("a\0b")
+  )
+
+  it("replace inserts NUL bytes from replacement (#1048)", ():
+    expect(replace("ab", "a", "x\0y")).to_eq("x\0yb")
+    expect(replace("aa", "a", "\0")).to_eq("\0\0")
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply tests/spec/strings.test.ry]: Fixed. Changed the 6 assertions whose expected value contains embedded `\0` to `expect(result == expected).to_eq(true)`, which routes through the NUL-safe `==` operator rather than `to_eq`'s string matcher. The two assertions in "replace finds and rewrites NUL needle" (`"a-b-a"`, `"XXX"`) produce NUL-free results and are unchanged.
